packages/server-wallet/src/protocols/__test__/process-ledger-queue.test.ts

```diff
@@ -0,0 +1,380 @@
+    fundingStrategy: 'Ledger',
+    vars: [
+      stateSignedBy([aliceSW(), bobSW()])({
+        turnNum: 1,
+        outcome: simpleEthAllocation(...allocations),
+      }),
+    ],
+  });
+
+const ledgerChannelWithAllocations = (...allocations: PreAllocationItem[]) =>
+  channel({
+    vars: [
+      stateSignedBy([aliceSW(), bobSW()])({
+        turnNum: 3,
+        outcome: simpleEthAllocation(...allocations),
+      }),
+    ],
+  }).protocolState as ChannelStateWithSupported;
+// END TEST HELPERS
+
+const defaultLedgerChannel = ledgerChannelWithAllocations([alice, 10], [bob, 10]);
+
+// const unfundedLedgerChannel = ledgerChannelWithAllocations([]);
+
+const processLedgerQueueProtocolState = (args: Partial<ProtocolState> = {}) => ({
+  fundingChannel: defaultLedgerChannel,
+
+  channelsRequestingFunds: [],
+  channelsReturningFunds: [],
+
+  theirLedgerProposal: undefined,
+  theirLedgerProposalNonce: 0,
+
+  myLedgerProposal: undefined,
+  myLedgerProposalNonce: 0,
+
+  ...args,
+});
+
+describe('marking ledger requests as complete', () => {
+  it('takes no action if there are no ledger requests', () => {
+    expect(protocol(processLedgerQueueProtocolState())).toBeUndefined();
+  });
+
+  it('detects completed funding requests from the outcome of supported state', () => {
+    const requestChannel = channel();
+    const protocolArgs = {
+      fundingChannel: ledgerChannelWithAllocations([requestChannel.channelId, 10]),
+      channelsRequestingFunds: [requestChannel.protocolState],
+    };
+    expect(protocol(processLedgerQueueProtocolState(protocolArgs))).toMatchObject({
+      type: 'MarkLedgerFundingRequestsAsComplete',
+      fundedChannels: [requestChannel.channelId],
+      defundedChannels: [],
+      ledgerChannelId: protocolArgs.fundingChannel.channelId,
+    });
+  });
+
+  it('detects completed defunding requests from the outcome of supported state', () => {
+    const requestChannel = channel();
+    const protocolArgs = {
+      fundingChannel: defaultLedgerChannel,
+      channelsReturningFunds: [requestChannel.protocolState],
+    };
+    expect(protocol(processLedgerQueueProtocolState(protocolArgs))).toMatchObject({
+      type: 'MarkLedgerFundingRequestsAsComplete',
+      fundedChannels: [],
+      defundedChannels: [requestChannel.channelId],
+      ledgerChannelId: protocolArgs.fundingChannel.channelId,
+    });
+  });
+});
+
+describe('exchanging ledger proposals', () => {
+  describe('as initial proposer', () => {
+    it('takes no action if proposal is identical to existing supported outcome', () => {
+      const requestChannel = prefundChannelWithAllocations(); // Empty outcome
+      const protocolArgs = {
+        fundingChannel: defaultLedgerChannel,
+        channelsRequestingFunds: [requestChannel.protocolState],
+      };
+      expect(protocol(processLedgerQueueProtocolState(protocolArgs))).toBeUndefined();
+    });
+
+    it('takes no action if proposal exists but counterparties does not', () => {
```

Comment:
Should it request a proposal from the peer in this case?

Reply:
The _wallet_ sends the proposal in this case, but the _protocol_ does not. The reason why is that there would need to be some state for the _protocol_ to be able to know "I have decided to request a proposal from my peer", so that the run loop can terminate.